federated_aws_cli/login.py

```diff
@@ -337,21 +340,29 @@ def aws_federate(self):
             "sessionToken": self.credentials["SessionToken"],
         }
 
-        params = urlencode({
+        query = urlencode({
             "Action": "getSigninToken",
             "Session": json.dumps(creds),
         })
+        logger.debug("Web Console params: {}".format(query))
 
-        logger.debug("Web Console params: {}".format(params))
-
-        url = "https://signin.aws.amazon.com/federation?Action=getSigninToken"
-        url += "&Session={}".format(quote_plus(json.dumps(creds)))
-
+        url_tuple = urlparse('https://signin.aws.amazon.com/federation')
+        url = urlunparse(url_tuple._replace(query=query))
         token = requests.get(url).json()
 
-        url = "https://signin.aws.amazon.com/federation?Action=login"
-        url += "&Destination=" + quote_plus("https://console.aws.amazon.com/")
-        url += "&SigninToken=" + token["SigninToken"]
+        account_id = self.role_arn.split(":")[4]
+        role = self.role_arn.split(':')[5].split('/')[-1]
+        issuer_url_query = urlencode({"account": account_id, "role": role})
+        issuer_url = urlunparse(
+            ('https', self.issuer_domain, '/', '', issuer_url_query, ''))
+
+        query = urlencode({
+            "Action": "login",
+            "Destination": "https://console.aws.amazon.com/",
+            "SigninToken": token["SigninToken"],
+            "Issuer": issuer_url
+        })
+        url = urlunparse(url_tuple._replace(query=query))
```

Comment:
What is `_.replace()`?

Reply:
[`_replace`](https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urlparse) is a method on a parsed URL which the docs recommend using to change values. Why it's a private method is beyond me
> As is the case with all named tuples, the subclass has a few additional methods and attributes that are particularly useful. One such method is _replace(). The _replace() method will return a new ParseResult object replacing specified fields with new values.